File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;
class Solution {
    public int solution(int[][] maps) {
        int[] dx = {-1,1,0,0};
        int[] dy = {0,0,-1,1};
        
        int answer = -1;
        int n = maps.length;
        int m = maps[0].length;
        Queue<int[]> q = new LinkedList<>();
        boolean[][] visited = new boolean[n][m];
        
        
        
        q.add(new int[]{0,0,1});
        visited[0][0] = true;
        
        while (!q.isEmpty()) {
            int[] node = q.poll();
            int x = node[0];
            int y = node[1];
            int cnt = node[2];
            
            if (x == n-1 && y == m-1) {
                return cnt;
            }
            
            for (int i=0;i<4;i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];
                
                if (nx >= 0 && nx < n && ny >= 0 && ny < m) {
                    if (!visited[nx][ny] && maps[nx][ny] == 1) {
                        q.add(new int[]{nx,ny,cnt+1});
                        visited[nx][ny] = true;
                    }
                }
            }
        }
        
        return answer;
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<bits/stdc++.h>
using namespace std;

queue<pair<int,int>> q;
int dy[] = {-1,1,0,0};
int dx[] = {0,0,-1,1};

int visited[104][104];
int y,x,nx,ny, n, m;

int solution(vector<vector<int> > maps)
{
    int answer = 0;

    n = maps.size();
    m = maps[0].size();

    q.push({0,0});
    visited[0][0] = 1;


    while (q.size()) {
        tie(y,x) =  q.front(); q.pop();

        for (int i=0;i<4;i++) {
            ny = y + dy[i];
            nx = x + dx[i];

            if (ny < 0 || ny >= n || nx < 0 || nx >= m) continue;
            if (maps[ny][nx] == 0 || visited[ny][nx] != 0) continue;
            visited[ny][nx] = visited[y][x] + 1;
            q.push({ny,nx});
        }
    }


    return visited[n-1][m-1] == 0 ? -1 : visited[n-1][m-1];
}